internal/route/routing_strategy.go

```diff
@@ -61,5 +75,86 @@ func (s *PriorityRoundRobinStrategy) RouteNextRequest(
 		s.logger.Debug("Did not find any healthy nodes in priority.", zap.Int("priority", priority))
 	}
 
+	// No healthy upstreams are available. If `alwaysRoute` is true, find an unhealthy upstream to route to anyway.
+	// TODO(polsar): At this time, the only unhealthy upstreams that can end up here are those due to high latency
+	//  or error rate. Pass ALL configured upstreams in `upstreamsByPriority`. Their health status should be indicated
+	//  in the UpstreamConfig.HealthStatus field.
+	if s.alwaysRoute {
```

Comment:
Let's try to do this now! I don't think it will be too bad. It could look something like this:

```
type NodeFilter interface {
	Apply(requestMetadata metadata.RequestMetadata, upstreamConfig *config.UpstreamConfig, numUpstreamsInPriorityGroup int) bool
        GetFilterType() NodeFilterType // New method on this interface
}

type AlwaysRouteRoutingStrategy struct {
	NodeFilters          []NodeFilter
        FiltersToRemove  []NodeFilterType // Order matters
	BackingStrategy  RoutingStrategy
	Logger                 *zap.Logger
}

func (s *AlwaysRouteRoutingStrategy) RouteNextRequest(
	upstreamsByPriority types.PriorityToUpstreamsMap,
	requestMetadata metadata.RequestMetadata,
) (string, error) {
	filters := s.NodeFilters
        filtersToRemove := s.FiltersToRemove

	for len(filters) > 0 {
		// Attempt to route with the current set of filters
		upstreams, err := s.filter(upstreamsByPriority, requestMetadata, filters)
		if err != nil {
			return "", err
		}

		if len(upstreams) > 0 {
			return s.BackingStrategy.RouteNextRequest(filteredUpstreams, requestMetadata)
		}

		// If no upstreams are found, remove the filter that's first in s.FiltersToRemove and try again
		filters = filters.remove(filterToRemove.getFirst()) // psuedocode
	}

	// If all filters are exhausted and no upstreams are found, pass in original list of upstreams
	return s.BackingStrategy.RouteNextRequest(upstreamsByPriority, requestMetadata)
}

// I just copied this from https://github.com/satsuma-xyz/node-gateway/blob/0e375d8e15f8c693e54be90c284f3ce29999e97f/internal/route/filtering_strategy.go#L24
func (s *AlwaysRouteRoutingStrategy) filter(
	upstreamsByPriority types.PriorityToUpstreamsMap,
	requestMetadata metadata.RequestMetadata,
	filters []NodeFilter,
) ([]string, error) {
	priorityToHealthyUpstreams := make(types.PriorityToUpstreamsMap)

	for priority, upstreamConfigs := range upstreamsByPriority {
		s.Logger.Debug("Determining healthy upstreams at priority.", zap.Int("priority", priority), zap.Any("upstreams", upstreamConfigs))

		filteredUpstreams := make([]*config.UpstreamConfig, 0)

		for _, upstreamConfig := range upstreamConfigs {
			ok := s.NodeFilter.Apply(requestMetadata, upstreamConfig, len(upstreamConfigs))
			if ok {
				filteredUpstreams = append(filteredUpstreams, upstreamConfig)
			}
		}

		priorityToHealthyUpstreams[priority] = filteredUpstreams
	}

	return priorityToHealthyUpstreams
}

```


Other benefits of this design:
- IsLatency/ErrorsAcceptable NodeFilter actual filters nodes instead of sets the UpstreamStatus in UpstreamConfig
- Removes the need to set/store an `HealthStatus` on the UpstreamConfig

Reply:
I discuss a slightly more optimized version of this routing algo [here](https://www.notion.so/alchemotion/Error-and-latency-based-routing-in-the-Satsuma-node-gateway-3a60d896c5844f82bf90f68570b00608?pvs=4#bf4dcee7ee1949a9bd960634c777751b). It's a bit more complex, though, so we don't have to do this now.

Reply:
Fair enough! Let's ship and get a round of real testing before addressing the followups.